template/base/src/env/client-env.mjs

```diff
@@ -0,0 +1,36 @@
+import { clientEnv, clientEnvSchema } from './env-schema.mjs';
+
+const _clientEnv = clientEnvSchema.safeParse(clientEnv)
+
+const formatErrors = (
+    /** @type {import('zod').ZodFormattedError<Map<string,string>,string>} */
+    errors,
+  ) =>
+    Object.entries(errors)
+      .map(([name, value]) => {
+        if (value && "_errors" in value)
+          return `${name}: ${value._errors.join(", ")}\n`;
+      })
+      .filter(Boolean);
+
+if (!_clientEnv.success) {
+  console.error(
+    "❌ Invalid environment variables:\n",
+    ...formatErrors(_clientEnv.error.format()),
+  );
+  throw new Error("Invalid environment variables");
+}
+   
+for (let key of Object.keys(_clientEnv.data)) {
+  if (!key.startsWith("NEXT_PUBLIC_")) {
```

Comment:
Add the "reverse" validation for server so that you can't "accidentally" expose secrets?

Reply:
Yea, also prevents typos and having to waste time figuring out why your public env vars aren't showing up in the browser.